go-controller/pkg/clustermanager/endpointslicemirror/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,428 @@
+package endpointslicemirror
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	nadlister "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/listers/k8s.cni.cncf.io/v1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	v1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	maxRetries = 10
+	// LabelSourceEndpointSlice label key used in mirrored EndpointSlice
+	// that has the value of the default EndpointSlice name
+	LabelSourceEndpointSlice = "k8s.ovn.org/source-endpointslice"
+	// LabelSourceEndpointSliceVersion label key used in mirrored EndpointSlice
+	// that has the value of the last known default EndpointSlice ResourceVersion
+	LabelSourceEndpointSliceVersion = "k8s.ovn.org/source-endpointslice-version"
+)
+
+// Controller represents the EndpointSlice mirror controller.
+// For namespaces that use a user-defined primary network, this controller mirrors the default EndpointSlices
+// (managed by the default Kubernetes EndpointSlice controller) into new EndpointSlices that contain the addresses
+// from the primary network.
+type Controller struct {
```

Comment:
Can we use the [`pkg/controller`](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/controller/controller.go#L120) ? it will remove a lot of k8s controllers infra.

Reply:
Unfortunately no, the endpointslicemirror controller has to have access to the objects that are being changed and this is not currently available through the controller abstraction. I can revisit that in the future, if time permits..

Reply:
You have this https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/controller/controller.go#L53

```golang
// ObjNeedsUpdate tells if object should be reconciled.
	// May be called with oldObj = nil on Add, won't be called on Delete.
	ObjNeedsUpdate func(oldObj, newObj *T) bool
}
```

It can be abused a little.

Reply:
maybe @npinaeva has some ideas here

Reply:
But from what I see on the PR we don't see the old version either 
https://github.com/ovn-org/ovn-kubernetes/pull/4499/files#diff-b57d5b4546060f8f01d74875fd7ef576e7730e05682891df27bda4045a40b045R95-R97

```golang
func (c *Controller) onEndpointSliceUpdate(_ interface{}, new interface{}) {
	c.enqueueEndpointSlice(new)
}
```

Reply:
Anyhow again, maybe we don't want to tackle this now, up to you @kyrtapz

Reply:
`enqueueEndpointSlice` calls `getDefaultEndpointSliceKey`. For mirrored endpointslices this functions extracts the default endpointslice key by looking for a label. There is no way to do that currently with the abstraction we have.
I am ok to revisit that as a followup but I don't see this as being necessary in this PR considering our tight timelines.

Reply:
fair enough, I will add a bullet to the tech debt card

Reply:
I am not sure I understand why the controller framework can't be used here, @qinqon suggestion makes sense to me. Feel free to ping me when we get back to this problem.

Reply:
We have discussed this offline and we will revisit this in the future as I cannot use the abstraction as is in this controller.